### ISSUE TITLE
Auth using Emulator with demo project

### ISSUE THREAD

**Issue by Ticlext-Altihaf** [kreait/firebase-php#844]
### Describe the feature you would like to see

Is it possible to use Firebase Emulator with demo Firebase project, it would be useful for CI testing where the current setup require creating new Firebase project and issue service account.

https://firebase.google.com/docs/emulator-suite/connect_auth#choose_a_firebase_project

Example from using official Admin SDK for Node.js, it works without requiring internet connection or having Firebase project.
`firebase emulators:start --project=demo-project`
```js
process.env.FIREBASE_AUTH_EMULATOR_HOST = "127.0.0.1:9099"
import { initializeApp } from 'firebase-admin/app';
import { getAuth } from 'firebase-admin/auth';

initializeApp({
    projectId: 'demo-project',
});

const newUserData = {
    email: Math.random()*10000 + '@example.com', // XXXX@example.com
    emailVerified: false,
    phoneNumber: '+1' + Math.random()*10000000000, // +1XXXXXXXXXX
    password: 'secretPassword',
    displayName: 'John Doe',
    photoURL: 'http://www.example.com/12345678/photo.png',
    disabled: false,
};

const user = await getAuth().createUser(newUserData)

// Getting the ID token with signInWithPassword
let idToken = await fetch(`http://127.0.0.1:9099/identitytoolkit.googleapis.com/v1/accounts:signInWithPassword?key=random`, {
    method: 'POST',
    headers: {
        'Content-Type': 'application/json',
    },
    body: JSON.stringify({
        email: user.email,
        password: "secretPassword",
        returnSecureToken: true,
    }),
})

idToken = await idToken.json()
idToken = idToken.idToken

const userFromIdToken = await getAuth().verifyIdToken(idToken)

await getAuth().deleteUser(userFromIdToken.uid)
```

**Comment by jeromegamez** [comment]
Yes, you can read about it here:

https://firebase-php.readthedocs.io/en/stable/testing.html#using-the-firebase-emulator-suite

**Comment by Ticlext-Altihaf** [comment]
Uh.. I think you misunderstood me, I did read the docs I can't find the guide where I can use Firebase Demo Project, yes it work with Real Firebase Project.

![image](https://github.com/kreait/firebase-php/assets/77704356/2c1d2da2-2f07-48a8-bf8c-92e21abafb8f)

The code, where in Node.js SDK it work without credentials.
```php
        $request = CreateUser::new()
            ->withEmail($email)
            ->withPhoneNumber($phoneNumber)
            ->withClearTextPassword($password)
            ->withDisplayName($faker->name);


        $firebaseUser = Firebase::auth()->createUser($request);
```
![image](https://github.com/kreait/firebase-php/assets/77704356/4f02cce8-6337-4631-96dd-e2e9252600f9)

**Comment by jeromegamez** [comment]
Ah, I understand now, sorry for closing the issue so early 🙏🏻 (it's early here 😅)

I think this should be working, I'm running emulator tests with it, but seeing the `demo-` prefix... I've perhaps been doing it wrong the whole time 😬

https://github.com/kreait/firebase-php/blob/393a751a0a15291774070688cbfa4d0e954b9e54/.github/workflows/emulator-tests.yml#L71

I'll try to look into it as soon as I find the time, I'm currently working on the next major release for PHP 8.3, and I do have only limited time...

**Comment by cmenkemeller** [comment]
@jeromegamez Did you ever find a solution for this? I am running into a similar issue where verifyIdToken() is making calls to google apis even though I'm pointing to the Firebase Emulators.

Offline fails with curl error - 
Next Kreait\Firebase\Exception\Database\ApiConnectionFailed: Unable to connect to the API: cURL error 6: Could not resolve host: oauth2.googleapis.com (see https://curl.haxx.se/libcurl/c/libcurl-errors.html) for https://oauth2.googleapis.com/token in ...vendor/kreait/firebase-php/src/Firebase/Exception/DatabaseApiExceptionConverter.php:37

**Comment by zackarydev** [comment]
When defining the env variable `"FIREBASE_AUTH_EMULATOR_HOST"` using dotenv, the Firebase Factory throws: `Unable to create an API client without credentials`

Is there anything missing @jeromegamez ?

```
$factory = (new Factory())
            ->withProjectId('demo-myProject')
            ->withDatabaseUri('http://localhost:9097');
```

When I edit the Factory createApiClient and remove the AuthTokenMiddleware "The request is missing a valid API key" is thrown.

From what I can tell, the `firebase-service-account.json` MUST be provided a real service account json file, to perform initial authorization of requests. But requests are indeed routed to the emulator afterwards.

**Comment by jeromegamez** [comment]
Yes, that's the current limitation and why this issue exists in the first place 😅.

The thing is that when I created the SDK, the emulator suite was far away in the future, and the current implementation (the SDK and the tests) doesn't leave much wiggle room.

The SDK has evolved since this issue was created, I'll see if I can make it work with the emulator ago, but I currently can't free up a few consecutive hours to look deeper into it. Demand for this is not very high (either people do have a real Firebase project for tests, or they wrap the SDK, so they don't have to invoke it during tests), and the lack of funding plays a part as well. 

If I do find the time, I'll put this higher on my bucket list, but you want to give it a try, I'd appreciate a PR.

**Comment by jeromegamez** [comment]
Of course I had to look into it after all and I might have found a way to improve the situation.

It needs some preparation, I'll create a PR and let you know so that you can test it.

**Comment by zackarydev** [comment]
> Demand for this is not very high (either people do have a real Firebase project for tests, or they wrap the SDK, so they don't have to invoke it during tests), and the lack of funding plays a part as well.

Yep that was the solution!

**Comment by jeromegamez** [comment]
As it turns out, my "I might have found a way to improve the situation" was too optimistic. It's as complicated as it was when the issue was created. This needs to wait until another time (or someone makes a PR, but this one's going to be something).
